graph/path/control_flow_bench_test.go

```diff
@@ -0,0 +1,272 @@
+	}{
+		{name: "gnm-n=1e3-m=1e3", g: gnm(1e3, 1e3)},
+		{name: "gnm-n=1e3-m=3e3", g: gnm(1e3, 3e3)},
+		{name: "gnm-n=1e3-m=1e4", g: gnm(1e3, 1e4)},
+		{name: "gnm-n=1e3-m=3e4", g: gnm(1e3, 3e4)},
+
+		{name: "gnm-n=1e4-m=1e4", g: gnm(1e4, 1e4)},
+		{name: "gnm-n=1e4-m=3e4", g: gnm(1e4, 3e4)},
+		{name: "gnm-n=1e4-m=1e5", g: gnm(1e4, 1e5)},
+		{name: "gnm-n=1e4-m=3e5", g: gnm(1e4, 3e5)},
+
+		{name: "gnm-n=1e5-m=1e5", g: gnm(1e5, 1e5)},
+		{name: "gnm-n=1e5-m=3e5", g: gnm(1e5, 3e5)},
+		{name: "gnm-n=1e5-m=1e6", g: gnm(1e5, 1e6)},
+		{name: "gnm-n=1e5-m=3e6", g: gnm(1e5, 3e6)},
+
+		{name: "gnm-n=1e6-m=1e6", g: gnm(1e6, 1e6)},
+		{name: "gnm-n=1e6-m=3e6", g: gnm(1e6, 3e6)},
+		{name: "gnm-n=1e6-m=1e7", g: gnm(1e6, 1e7)},
+		{name: "gnm-n=1e6-m=3e7", g: gnm(1e6, 3e7)},
+
+		{name: "dup-n=1e3-d=0.8-a=0.1", g: duplication(1e3, 0.8, 0.1, math.NaN())},
+		{name: "dup-n=1e3-d=0.5-a=0.2", g: duplication(1e3, 0.5, 0.2, math.NaN())},
+
+		{name: "dup-n=1e4-d=0.8-a=0.1", g: duplication(1e4, 0.8, 0.1, math.NaN())},
+		{name: "dup-n=1e4-d=0.5-a=0.2", g: duplication(1e4, 0.5, 0.2, math.NaN())},
+
+		{name: "dup-n=1e5-d=0.8-a=0.1", g: duplication(1e5, 0.8, 0.1, math.NaN())},
+		{name: "dup-n=1e5-d=0.5-a=0.2", g: duplication(1e5, 0.5, 0.2, math.NaN())},
+	}
+
+	for _, test := range tests {
+		rnd := rand.New(rand.NewSource(1))
+		g := test.g()
+
+		// Guess a maximally expensive entry to the graph.
+		sort, err := topo.Sort(g)
+		root := sort[0]
+		if root == nil {
+			// If we did not get a node in the first position
+			// then there must be an unorderable set of nodes
+			// in the first position of the error. Pick one
+			// of the nodes at random.
+			unordered := err.(topo.Unorderable)
+			root = unordered[0][rnd.Intn(len(unordered[0]))]
+		}
+		if root == nil {
+			b.Error("no entry node label for graph")
+			continue
+		}
+
+		if len(sort) > 1 {
+			// Ensure that the graph has a complete path
+			// through the sorted nodes.
+
+			// unordered will only be accessed if there is
+			// a sort element that is nil, in which case
+			// unordered will contain a set of nodes from
+			// an SCC.
+			unordered, _ := err.(topo.Unorderable)
+
+			var ui int
+			for i, v := range sort[1:] {
+				u := sort[i]
+				if u == nil {
+					u = unordered[ui][rnd.Intn(len(unordered[ui]))]
+					ui++
+				}
+				if v == nil {
+					v = unordered[ui][rnd.Intn(len(unordered[ui]))]
+				}
+				if !g.HasEdgeFromTo(u, v) {
+					g.SetEdge(g.NewEdge(u, v))
+				}
+			}
+		}
+
+		b.Run(test.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				d := Dominators(root, g)
+				if got := d.Root(); got.ID() != root.ID() {
+					b.Fatalf("unexpected root node: got:%d want:%d", got.ID(), root.ID())
+				}
+			}
+		})
+	}
+}
+
+// gnm returns a directed G(n,m) Erdõs-Rényi graph.
+func gnm(n, m int) func() *simple.DirectedGraph {
+	return func() *simple.DirectedGraph {
+		dg := simple.NewDirectedGraph()
+		err := gen.Gnm(dg, n, m, rand.New(rand.NewSource(1)))
```

Comment:
Given the same random source `rand.NewSource(1)` would `gen.Gnm` always return the same graph, given the same `n` and `m`?

If so, is this intended for the returned closure which generates random graphs given `n` and `m`? Several invokations to the closure would then results in duplicate graphs, so the test would run on the same graph, rather than distinct graphs for a given `n` and `m`.

I may have misunderstood some part, if so, feel free to fill in my gaps in understanding.